pkg/services/meta/notifications.go

```diff
@@ -0,0 +1,497 @@
+	}
+
+	return nil
+}
+
+func (m *Meta) listenNotifications(ctx context.Context) error {
+	for {
+		select {
+		case h := <-m.bCh:
+			go func() {
+				err := m.handleBlock(h.Index)
+				if err != nil {
+					m.l.Error(fmt.Sprintf("processing %d block", h.Index), zap.Error(err))
+					return
+				}
+			}()
+		case aer := <-m.objEv:
+			// TODO: https://github.com/nspcc-dev/neo-go/issues/3779 receive somehow notifications from blocks
+
+			l := m.l.With(zap.Stringer("notification container", aer.Container))
+
+			switch aer.Name {
+			case objPutEvName:
+				ev, err := parseObjNotification(aer)
+				if err != nil {
+					l.Error("invalid object notification received", zap.Error(err))
+					continue
+				}
+
+				m.m.RLock()
+				_, ok := m.storages[ev.cID]
+				m.m.RUnlock()
+				if !ok {
+					l.Debug("skipping object notification", zap.Stringer("inactual container", ev.cID))
+					continue
+				}
+
+				go func() {
+					err = m.handleObjectNotification(ev)
+					if err != nil {
+						l.Error("handling object notification", zap.Error(err))
+						return
+					}
+				}()
+			case cnrDeleteName:
+				ev, err := parseCnrNotification(aer)
+				if err != nil {
+					l.Error("invalid container notification received", zap.Error(err))
+					continue
+				}
+
+				m.m.RLock()
+				_, ok := m.storages[ev.cID]
+				m.m.RUnlock()
+				if !ok {
+					l.Debug("skipping container notification", zap.Stringer("inactual container", ev.cID))
+					continue
+				}
+
+				go func() {
+					err = m.dropContainer(ev.cID)
+					if err != nil {
+						l.Error("deleting container failed", zap.Error(err))
+						return
+					}
+
+					l.Debug("deleted container", zap.Stringer("cID", ev.cID))
+				}()
+			case newEpochName:
+				epoch, err := parseEpochNotification(aer)
+				if err != nil {
+					l.Error("invalid new epoch notification received", zap.Error(err))
+					continue
+				}
+
+				go func() {
+					err = m.handleEpochNotification(epoch)
+					if err != nil {
+						l.Error("handling new epoch notification", zap.Int64("epoch", epoch), zap.Error(err))
+						return
+					}
+				}()
+			default:
+				m.l.Warn("unexpected notification", zap.String("event name", aer.Name))
+				continue
+			}
+		case <-ctx.Done():
+			m.l.Info("stop listening meta notifications")
+			return nil
+		}
+	}
+}
+
+const (
+	gcInterval    = 100
+	collapseDepth = 10
+)
+
+func (m *Meta) handleBlock(ind uint32) error {
+	m.m.RLock()
+	defer m.m.RUnlock()
+
+	for _, st := range m.storages {
+		// TODO: parallelize depending on what can parallelize well
+
+		st.m.Lock()
+
+		p := st.path
+		root := st.mpt.StateRoot()
+		if st.opsBatch != nil {
+			_, err := st.mpt.PutBatch(mpt.MapToMPTBatch(st.opsBatch))
+			if err != nil {
+				st.m.Unlock()
+
+				return fmt.Errorf("put batch for %d block to %q storage: %w", ind, p, err)
+			}
+
+			st.opsBatch = nil
+		}
+
+		st.m.Unlock()
+
+		st.mpt.Flush(ind)
+		st.mpt.Collapse(collapseDepth)
```

Comment:
Collapsing for every block is not very efficient.

Reply:
can you suggest a rule? connect it to real-time? 30 seconds? minutes?

Reply:
No hard rules here. It needs to be collapsed each time you think it's too big. Go figure.